README.txt
JSON-RPC 2.0 Client

Copyright (c) Vladimir Dzhuvinov, 2010 - 2012


README

Provides a client-side class for dispatching requests and notifications to a 
JSON-RPC 2.0 server. The JSON-RPC 2.0 messages are sent by HTTP(S) POST.

Optional client session features:

	* Customise the "Content-Type" header in HTTP POST requests.
	
	* Set an "Origin" header in HTTP POST requests to simulate Cross-Origin 
	  Resource Sharing (CORS) requests from a browser.
	  
	* Accept HTTP cookies (if client sessions are established by this mean
	  instead of through the JSON-RPC protocol itself).
	  
	* Customise the allowable "Content-Type" header values in HTTP POST 
	  responses.
	  
	* Preserve parse order of JSON object members in JSON-RPC 2.0 response 
	  results (for human facing clients, e.g. the JSON-RPC 2.0 Shell).
	  
	* Ignore version 2.0 checks when parsing responses to allow client 
	  sessions to older JSON-RPC (1.0) servers.

	* Parse non-standard attributes appended in JSON-RPC 2.0 responses.
	
	* Trust all X.509 server certificates (for HTTPS connections), including
	  self-signed.



Requirements:

	* Java 1.5 or later
	
	* The JSON-RPC 2.0 Base package com.thetransactioncompany.jsonrpc2 to 
	  represent, serialise and parse JSON-RPC 2.0 messages. Available from
	  http://software.dzhuvinov.com/json-rpc-2.0-base.html



Package content:

        README.txt                    This file.
        
        LICENSE.txt                   The software license.
	
	CHANGELOG.txt                 The change log.
	
	jsonrpc2-client-{version}.jar JAR file containing the compiled package
	                              classes.
				  
	javadoc/                      The Java Docs for this package.
	
	Example.java                  Example client session usage.
	
	build.xml                     The Apache Ant build file.
	
	lib/                          The package dependencies and their licenses.
        
	src/			      The source code for this package.
	
	test/                         JUnit tests for this package.



For complete JSON-RPC 2.0 Client documentation, examples and updates visit
	
	http://software.dzhuvinov.com/json-rpc-2.0-client.html


The JSON-RPC 2.0 specification and user group forum can be found at

	http://groups.google.com/group/json-rpc


[EOF]


CHANGELOG.txt
CHANGE LOG

JSON-RPC 2.0 Base, Copyright (c) Vladimir Dzhuvinov, 2009 - 2012


version 1.0 (2009-10-04)
	* First release.

version 1.1 (2009-12-04)
	* Official release.
	* Major refactoring.

version 1.2 (2009-12-27)
	* More detailed messages in parser exceptions.

version 1.3 (2010-01-06)
	* Adds convenient constant instances of the JSONRPC2Error class for all 
	  standard JSON-RPC 2.0 errors (with proper codes and messages): 
	  PARSE_ERROR, INVALID_REQUEST, METHOD_NOT_FOUND, INVALID_PARAMS, 
	  INTERNAL_ERROR.

version 1.4 (2010-02-02)
	* New JSONRPC2ParamsType class enumerates the three possible JSON-RPC 
	  parameter types: NO_PARAMS, ARRAY and OBJECT.
	* Improves the documentation.
	
version 1.5 (2010-02-12)
	* Adds the utility classes PositionalParamsRetriever and 
	  NamedParamsRetriever to ease the processing of request parameters 
	  (type and mandatory/optional parameter checking, default values).
	* Introduces JUnit test cases.

version 1.6 (2010-02-18)
	* Adds a base absract class ParamsRetriever to the utility parameter 
	  retriever classes PositionalParamsRetriever and NamedParamsRetriever.
	* Adds methods for getting enumerated strings to the parameter retriever 
	  classes.
	  
version 1.7 (2010-02-24)
	* Adds a getNames() method to NamedParamsRetriever to list the names of 
	  all parameters.
	* Adds the checking methods checkNames(String[] mandatoryNames) and 
	  checkNames(String[] mandatoryNames,String[]optionalNames) to 
	  NamedParamsRetriever which will throw an exception if a mandatory 
	  parameter name is missing or a name outside the expected is present.
	  
version 1.8 (2010-03-11)
	* Changes the signature of PositionalParamsRetriever.getOptEnumString().
	* Changes the signature of NamedParamsRetriever.getOptEnumString().

version 1.9 (2010-03-12)
	* Refactors the PositionalParamsRetriever class, hasParameter() and 
	  ensureParameter() methods.
	* Refactors the NamedParamsRetriever class, hasParameter() and 
	  ensureParameter() methods.

version 1.9.1 (2010-03-15)
	* Minor internal refactoring in the util classes.

version 1.9.2 (2010-08-07)
	* Util parameter retriever methods getEnumString() add support for Java 
	  enum constants.

version 1.9.3 (2010-08-09)
	* Adds explicit class parameter to util retriever methods getOpt().
	
version 1.9.4 (2010-08-19)
	* Renames util PositionalParamsRetriever.getEnum()|getOptEnum() and
	  NamedParamsRetriever.getEnum()|getOptEnum() methods

version 1.10 (2010-09-07)
	* Adds util retriever methods to allow optional null values for
	  JSON string, JSON array and JSON object parameters. The generic
	  ensure(), get() and getOpt() methods are also updated for this.

version 1.11 (2010-10-04)
	* Factors out common parse logic to the JSONRPC2Message class.
	* Parse exceptions provide specific message on empty JSON string.

version 1.12 (2010-12-12)
	* Released under the Apache Open Source License 2.0.

version 1.13 (2011-03-08)
	* Minor changes to the Ant build.xml script.
	* Includes version number in the JAR file.

version 1.14 (2011-03-09)
	* Allows optional parsing of JSON objects where their original member
	  order is preserved.
	* Allows for non-strict parsing of JSON-RPC 2.0 messages where the
	  "jsonrpc" version field is not checked.
	* Fixes JSONArray cast bug in JSONRPC2Message.parseJSONObject().
	* Documentation improvements.

version 1.15 (2011-05-19)
	* Allows for concurrent execution of the static JSONRPC2Message.parse
	  method.
	* Includes field in JSONRPCParseException to indicate the cause as
	  bad JSON or protocol exception.

version 1.16 (2011-05-19)
	* Refactors parsing, adds a reusable JSONRPC2Parser class.

version 1.17 (2011-06-11)
	* Switches the underlying JSON library from JSON.simple to the interface
	  compatible JSON Smart 1.0.6.4 for faster and more efficient parsing.
	* Expands use of Java generics.

version 1.18 (2011-07-07)
	* Completes the introduction of Java generics into the utility retriever
	  classes.

version 1.19 (2011-07-10)
	* Renames parser methods "noStrict" to "ignoreversion" and "isNoStrict" 
	  to "ignoresversion".
	* Adds methods for appending non-standard attributes (as aditional JSON 
	  object members) to JSON-RPC 2.0 messages.
	* Additional parser constructors and static parse methods to allow
	  setting of non-standard attribute parsing.

version 1.20 (2011-07-13)
	* Re-introduce parser methods "noStrict" and "isNoStrict" as deprecated.
	* Documentation improvements.
	* Updates JSON Smart JAR to 1.0.9.

version 1.21 (2011-08-05)
	* Provides an equals method implementation to the JSONRPC2Error class
	  which compares the error code values.

version 1.21.1 (2011-08-06)
	* Replaces String.isEmpty() by String.length()==0 for Google Android
	  compatibility (pre 9 level APIs).

version 1.22 (2011-10-08)
	* Utility parameter retriever classes relax the requirement of the
	  underlying datatype for int, long, float and double values by
	  converting from the more general java.lang.Number representation.
	* Removes the get{Opt}[Type]Array(...) methods from the utility 
	  parameter retriever classes.

version 1.23 (2011-10-09)
	* Reintroduces the previously removed utility parameter retriever 
	  methods get{Opt}StringArray(...).

version 1.24 (2011-10-25)
	* Adds methods hasParameters(String[] names) and 
	  hasParameters(String[] mandatoryNames, String[] optionalNames) to the
	  utility NamedParamsRetriever class.
	* Updates JSON Smart JAR to 1.0.9-1 (fixes parse bug #15).

version 1.25 (2012-02-19)
	* Updates JSON Smart JAR to 1.1.

version 1.25.1 (2012-04-03)
	* Updades JSON Smart JAR to 1.1.1.

version 1.26 (2012-07-08)
	* Adds class cast exception checking to critical sections in the utility
	  retriever classes.

version 1.26.1 (2012-07-15)
	* Add pom.xml.


src/main/java/com/thetransactioncompany/jsonrpc2/JSONRPC2Error.java
package com.thetransactioncompany.jsonrpc2;

import net.minidev.json.JSONObject;


/** 
 * Represents a JSON-RPC 2.0 error that occured during the processing of a 
 * request.
 *
 * <p>The protocol expects error objects to be structured like this:
 *
 * <ul>
 *     <li>{@code code} An integer that indicates the error type.
 *     <li>{@code message} A string providing a short description of the 
 *         error. The message should be limited to a concise single sentence.
 *     <li>{@code data} Additional information, which may be omitted. Its 
 *         contents is entirely defined by the application.
 * </ul>
 * 
 * <p>Note that the "Error" word in the class name was put there solely to
 * comply with the parlance of the JSON-RPC spec. This class doesn't inherit 
 * from {@code java.lang.Error}. It's a regular subclass of 
 * {@code java.lang.Exception} and, if thrown, it's to indicate a condition 
 * that a reasonable application might want to catch.
 *
 * <p>This class also includes convenient final static instances for all 
 * standard JSON-RPC 2.0 errors:
 *
 * <ul>
 *     <li>{@link #PARSE_ERROR} JSON parse error (-32700)
 *     <li>{@link #INVALID_REQUEST} Invalid JSON-RPC 2.0 Request (-32600)
 *     <li>{@link #METHOD_NOT_FOUND} Method not found (-32601)
 *     <li>{@link #INVALID_PARAMS} Invalid parameters (-32602)
 *     <li>{@link #INTERNAL_ERROR} Internal error (-32603)
 * </ul>
 *
 * <p>Note that the range -32099..-32000 is reserved for additional server 
 * errors.
 *
 * <p id="map">The mapping between JSON and Java entities (as defined by the 
 * underlying JSON Smart library): 
 * <pre>
 *     true|false  <--->  java.lang.Boolean
 *     number      <--->  java.lang.Number
 *     string      <--->  java.lang.String
 *     array       <--->  java.util.List
 *     object      <--->  java.util.Map
 *     null        <--->  null
 * </pre>
 *
 * <p>The JSON-RPC 2.0 specification and user group forum can be found 
 * <a href="http://groups.google.com/group/json-rpc">here</a>.
 *
 * @author Vladimir Dzhuvinov
 * @version 1.26 (2011-08-05)
 */
public class JSONRPC2Error extends Exception {
	
	
	/** 
	 * JSON parse error (-32700).
	 */
	public static final JSONRPC2Error PARSE_ERROR = new JSONRPC2Error(-32700, "JSON parse error");
	
	
	/** 
	 * Invalid JSON-RPC 2.0 request error (-32600).
	 */
	public static final JSONRPC2Error INVALID_REQUEST = new JSONRPC2Error(-32600, "Invalid request");
	
	
	/** 
	 * Method not found error (-32601). 
	 */
	public static final JSONRPC2Error METHOD_NOT_FOUND = new JSONRPC2Error(-32601, "Method not found");
	
	
	/** 
	 * Invalid parameters error (-32602).
	 */
	public static final JSONRPC2Error INVALID_PARAMS = new JSONRPC2Error(-32602, "Invalid parameters");
	
	
	/** 
	 * Internal JSON-RPC 2.0 error (-32603).
	 */
	public static final JSONRPC2Error INTERNAL_ERROR = new JSONRPC2Error(-32603, "Internal error");
	
	
	/**
	 * The error code.
	 */
	private int code;
	
	
	/**
	 * The optional error data.
	 */
	private Object data;
	
	
	/** 
	 * Creates a new JSON-RPC 2.0 error with the specified code and 
	 * message. The optional data is omitted.
	 * 
	 * @param code    The error code (standard pre-defined or
	 *                application-specific).
	 * @param message The error message.
	 */
	public JSONRPC2Error(int code, String message) {
		
		super(message);
		this.code = code;
	}
	
	
	/** 
	 * Creates a new JSON-RPC 2.0 error with the specified code,
	 * message and data.
	 * 
	 * @param code    The error code (standard pre-defined or
	 *                application-specific).
	 * @param message The error message.
	 * @param data    Optional error data, must <a href="#map">map</a>
	 *                to a valid JSON type.
	 */
	public JSONRPC2Error(int code, String message, Object data) {
		
		super(message);
		this.code = code;
		this.data = data;
	}
	
	
	/** 
	 * Gets the JSON-RPC 2.0 error code.
	 *
	 * @return The error code.
	 */
	public int getCode() {
		
		return code;
	}
	
	
	/**
	 * Gets the JSON-RPC 2.0 error data.
	 *
	 * @return The error data, {@code null} if none was specified.
	 */
	public Object getData() {
		
		return data;	
	}
	
	
	/** 
	 * Gets a JSON representation of the JSON-RPC 2.0 error.
	 *
	 * @return A JSON object representing this error object.
	 */
	public JSONObject toJSON() {
	
		JSONObject out = new JSONObject();
		
		out.put("code", code);
		out.put("message", super.getMessage());
		if (data != null)
			out.put("data", data);
				
		return out;
	}
	
	
	/** 
	 * Serialises the error object to a JSON string.
	 *
	 * @return A JSON-encoded string representing this error object.
	 */
	public String toString() {
		
		return toJSON().toString();
	}
	
	
	/**
         * Overrides {@code Object.equals()}.
         *
         * @param object The object to compare to.
         *
         * @return {@code true} if both objects are instances if this class and
	 *         their error codes are identical, {@code false} if not.
         */
        public boolean equals(Object object) {
        
                return object instanceof JSONRPC2Error && code == ((JSONRPC2Error)object).getCode();
        }
}
